Add DH, DSA, ECDH and ECDSA support in PKA Engine.

mahantesh-nvidia opened the pull request.

kblaiech commented on src/engine/helper/ec_local.h:
```diff
@@ -0,0 +1,295 @@
+#include <openssl/ec.h>
+#include <openssl/bn.h>
+
+/* Note: Although these structures are internal to OpenSSL, they are defined
```
multi-line comments should be as follow:

     /*
      * Multi-line
      * comment.
      */

mahantesh-nvidia replied:
Yes, Thanks.

kblaiech commented on src/engine/helper/ec_local.h:
```diff
@@ -0,0 +1,295 @@
+ *       API's the fields inside the structure have to be changed directly as
+ *       OpenSSL doesn't provide API's to do so.
+ *
+ *       Care needs to be taken to preserve the order and members of the structure
```
Good point! We have to keep it consistent across supported OpenSSL libraries

kblaiech commented on src/engine/helper/ec_local.h:
```diff
@@ -0,0 +1,295 @@
+ *
+ *       Structures are mirrored from crypto/ec/ec_local.h inside openssl github repo.
+ *       Prior to openssl 1.1.1 version, this file is named as ec_lcl.h
+ */
```
I think this not should be at the top, prior to the copyright:

  /*
   * Note goes here.
   *
   * Copyright goes here.
   *
   * License goes here.
   */

mahantesh-nvidia replied:
Ack, Thanks.

kblaiech commented on src/engine/helper/ec_local.h:
```diff
@@ -0,0 +1,295 @@
+ *       same as in the respective versions of OpenSSL.
+ *
+ *       Structures are mirrored from crypto/ec/ec_local.h inside openssl github repo.
+ *       Prior to openssl 1.1.1 version, this file is named as ec_lcl.h
```
Is it an exact copy of the file ec_lcl.h, or only used structures are copied?

mahantesh-nvidia replied:
Only the used structures are copied. There are many other definitions which require inclusion of other internal headers. These other definitions are not useful to us. Hence, only used structures are copied.

kblaiech commented on src/engine/helper/pka_helper.c:
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
+    operand_byte_copy(operand, (uint8_t *) bn_buf_ptr, buf_len);
+
+}
+static void bignum_to_ecc_operand(pka_bignum_t  *bignum,
```
Please leave an empty line

mahantesh-nvidia replied:
Will do, Thanks.

kblaiech commented on src/engine/helper/pka_helper.c:
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
+    if (operand == NULL)
+        return;
+
+    uint32_t byte_len, byte_max_len;
```
Must be at the top. Please do not mix code and variable declaration.

mahantesh-nvidia replied:
Ack, Thanks.

kblaiech commented on src/engine/helper/pka_helper.c:
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
     free(operand);
 }
 
+/* ECC related helpers */
```
coding style issue: I think in this file comments are as follow:

  // One line comment

  //
  // Multi-line comment
  //

mahantesh-nvidia replied:
Ack, Thanks.

kblaiech commented on src/engine/helper/pka_helper.c:
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
+    return;
+}
+
+static void malloc_ecc_operand(uint32_t       buf_len,
```
Can't we use malloc_operand() as defined in pka_helper.c?
I think this is redundant code.
same for the functions below?

mahantesh-nvidia replied:
As `ecc_point_t` has x, y variables of type `pka_operand_t` and not `pka_operand_t*`, we don't dynamically allocate and hence free this memory. As this `malloc()` is done for operand in` malloc_operand()`, we cannot use the same here. As discussed, I'll add "_buf" suffix to the function name as the current name is misleading.

kblaiech commented on src/engine/helper/pka_helper.c:
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
+    return point;
+}
+
+static void free_ecc_operand(pka_operand_t *operand)
```
same here. I think we can re-use free_operand() function instead

kblaiech commented on src/engine/helper/pka_helper.c:
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
+    buf_ptr = operand->buf_ptr;
+
+    if (buf_ptr != NULL)
+        free(buf_ptr);
```
free(operand)?

mahantesh-nvidia replied:
`ecc_point_t `has x,y variables declared as `pka_operand_t`, hence these operands are not dynamically created.

kblaiech commented on src/engine/helper/pka_helper.c:
```diff
@@ -629,6 +837,143 @@ int pka_rsa_mod_exp_crt(pka_bignum_t  *bn_value,
+    bignum_to_ecc_operand(bn_y, &(pt.y));
+    mltpr  = bignum_to_operand(bn_multiplier);
+
+    result = pka_do_ecc_pt_mult(tls_handle, &curve, &pt, mltpr);
```
To be consistent with the rest of the function calls, I think it is better to declare pka_do_ecc_pt_mult as:

   int pka_do_ecc_pt_mult(..., ecc_point *result)

mahantesh-nvidia replied:
As discussed, the current declaration is consistent with the rest of function calls. But I'll initialize the `rc` variable at the declaration time.

kblaiech commented on src/engine/helper/pka_helper.h:
```diff
@@ -112,6 +112,35 @@ int pka_rsa_mod_exp_crt(pka_bignum_t *bn_value,
 
+// This functions implements the elliptic curve point addition using
+// Bluefield PKA hardware.
+int pka_ecc_bn_pt_add(pka_bignum_t *bn_p,
```
pka_bn_ecc_pt_add() to be consistent with the above.

mahantesh-nvidia replied:
Ack, Thanks.

kblaiech commented on src/engine/helper/pka_helper.h:
```diff
@@ -112,6 +112,35 @@ int pka_rsa_mod_exp_crt(pka_bignum_t *bn_value,
+
+// This functions implements the elliptic curve point multiplication using
+// Bluefield PKA hardware.
+int pka_ecc_bn_pt_mult(pka_bignum_t *bn_p,
```
Same here

mahantesh-nvidia replied:
Ack, Thanks.

kblaiech commented on src/engine/helper/pka_helper.h:
```diff
@@ -112,6 +112,35 @@ int pka_rsa_mod_exp_crt(pka_bignum_t *bn_value,
+                      pka_bignum_t *bn_result_x,
+                      pka_bignum_t *bn_result_y);
+
+// This functions implements the elliptic curve point multiplication using
```
typo. function instead of functions

mahantesh-nvidia replied:
Ack, Thanks.

kblaiech commented on src/engine/helper/pka_helper.h:
```diff
@@ -112,6 +112,35 @@ int pka_rsa_mod_exp_crt(pka_bignum_t *bn_value,
                         pka_bignum_t *bn_qinv,
                         pka_bignum_t *bn_result);
 
+// This functions implements the elliptic curve point addition using
```
typo. function instead of functions

kblaiech commented on src/engine/helper/pka_helper.h:
```diff
@@ -112,6 +112,35 @@ int pka_rsa_mod_exp_crt(pka_bignum_t *bn_value,
+
+// This function implements the modular inverse using BlueField
+// PKA hardware.
+int pka_mod_inv(pka_bignum_t *bn_value,
```
pka_bn_mod_inv?

mahantesh-nvidia replied:
Yes, correct. Thanks.

kblaiech commented on src/lib/pka_queue.c:
```diff
@@ -155,6 +155,33 @@ static uint16_t pka_concat6_wlen(uint32_t word_len,
+// Adjustment is done by removing zero padding at the most significant bytes.
+// Once this is done, align the new length around 8-byte boundary.
+// Note: Here it's assumed that the value is in Little Endian format.
+static void pka_adjust_rslt(pka_operand_t *result)
```
I believe "adjust" is a bit confusing. Could you please call it something like: pka_queue_no_zero_pad_rslt or any other name that explicitly indicates the type of the operation?

Note that, by convention, all functions in pka_queue.c have the prefix pka_queue_*

mahantesh-nvidia replied:
Yes, I'll replace "adjust" with something more relevant. However, as discussed, the inclusion of "queue" in the function name is not necessary as this is a static function. Thanks

kblaiech commented on src/lib/pka_queue.c:
```diff
@@ -155,6 +155,33 @@ static uint16_t pka_concat6_wlen(uint32_t word_len,
+    byte_ptr    = result->buf_ptr;
+
+    // Result from pka hardware is always in Little Endian.
+    byte_ptr = &byte_ptr[result2_len - 1];
```
This is true. But is it always true? what if the hardware change? Can we consider an if statement that handles both BE and LE. I think the byte order is stored somewhere, thus we can easily retrieve the information and apply the right logic.

mahantesh-nvidia replied:
I'll add the BE logic, Thanks.

mahantesh-nvidia commented:
> Looks good to me except few comments I would like you to address and other minor issues related to the coding style.

Thanks for the review @kblaiech , I have addressed your review comments.
